### ISSUE TITLE
Distribution.Simple.Utils.writeFileAtomic doesn't handle mode and ownership correctly

### ISSUE THREAD

**Issue by bgamari** [haskell/cabal#7642]
**Describe the bug**
While looking at [GHC #14017](https://gitlab.haskell.org/ghc/ghc/-/issues/14017) I noticed that GHC's implementation of `writeFileAtomic` (which is apparently copied from Cabal's `Distribution.Simple.Utils`) does not handle permissions correct. That is, the mode and owner of the overwritten file will not be preserved during the update. This could break, especially when umask is overridden

**To Reproduce**

```
$ touch hi
$ chmod 0077 hi
$ chown :dialout hi
$ ghci -package Cabal
> writeFileAtomic "hi" mempty
> :quit
$ ls -lh hi
-rw-r--r-- 1 ben users 0 Sep 13 15:45 hi
```

**Expected behavior**
The mode and owner of the file `hi` should be unchanged.

**System information**
 - Operating system
 - `cabal`, `ghc` versions

**Additional context**

**Comment by DanielG** [comment]
What makes you think this function should preserve the perm bits? Seems reasonable to me to reset them given that it has to remove the original file in order to be atomic.

If an admin want to change the mode/owner of files in the way you demonstrate they should use the setgid bit on the containing directory tree instead, which works now after #7572 :)

**Comment by DanielG** [comment]
I've posted what I think is a more appropriate [fix for the GHC issue (MR#6558)](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6558] haven't tested it yet though.

**Comment by DanielG** [comment]
I've done a quick review of all the writeFileAtomic[1] call sites to check if we have any spots where the written files would be expected to be world readable and we have to `chmod ugo+r` like we do in setFileOrdinary and friends.

[1] wrappers around writeFileAtomic include: writeUTF8File, rewriteFileLBS, writeGenericPackageDescription

Notable examples of files written with that function but without a chmod call:

writeFileAtomic:

- GHC environment files in the project directory in writeGhcEnvironmentFile and in cabal-install installLibraries,
-  The persistent build config aka. LoclaBuildInfo aka.`dist/setup-config` in writePersistBuildConfig, 
-  (I think) 0x-index.tar in updateRepo,
- source tarball written by `cabal get` in unpackPackage
- updated .cabal file in build directory in installUnpackedPackage from cabal-install
- FileMonitor cache files in rewriteCacheFile
- downloaded files in plainHttpTransport

I would say that as long as the written files stay in a build directory and aren't installed into a destdir/prefix them not being world readable is fine and actually what you'd expect. Call sites for the writeUTF8File wrapper seem more dangerous though:

- writeRegistrationFileDirectly, HcPkg.hs
- maybeCreateDefaultSetupScript and overwriteSnapshotPackageDesc in SrcDist.hs (probably get tar'ed up into the srcdist so these could be problematic)
- registerAll in Register.hs (involved in installation so needs review)

**Comment by Mikolaj** [comment]
Ping? A very narrow window to get this into cabal 3.8 (if anything needs to be done).

**Comment by DanielG** [comment]
If the bits above weren't causing problems before it's probably fine but I haven't had time to do any more review here.

**Comment by DanielG** [comment]
Just to clarify on the GHC side my [!6558](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6558) was merged, mirroring the behaviour established by the fix(es) in cabal for #7560 and so the behavior Ben was proposing here isn't what we went with in the end.

**Comment by Mikolaj** [comment]
In that case, not to leave a mildly misleading ticket alive, should we close it and open a new one if anything turns out amiss?
